### PR TITLE
fix: correct retry logic

### LINKED ISSUE #12543: Retrying specific failed node does not work
# Suggested Enhancement

Allow users to selectively retry specific failed nodes instead of retrying all failed nodes at once.

## Use Cases

I'm using Argo Workflow, and at times, I would like the option to retry a specific failed node, instead of retrying all failed nodes (similar to GitLab CI's capability). Even if the overall pipeline still ends up failing, there are specific tasks that I'd prefer to retry without consuming excessive resources retrying other nodes that may inevitably fail. I believe providing users with this level of flexibility is important.

For example, in the following pipeline, I might prefer to only rerun the failed nodes of _BB_, rather than retrying both _B_ and _C_ nodes.

image

*--- updates by agilgur5 below---*
- `--node-field-selector` already exists and [can be used](https://github.com/argoproj/argo-workflows/issues/12543#issuecomment-1898832839) to specify nodes
- but there appears to be [a bug for failed nodes](https://github.com/argoproj/argo-workflows/issues/12543#issuecomment-1899798080), so re-labeled and re-titled this as a bug issue

### PR DESCRIPTION
**Pull request by isubasinghe**
Fixes #12543 and other retry related issues.

### Motivation
The current retry logic is too simplistic, it relies only on resetting nodes by traversing boundary nodes.
This approach does not work, it needs to be a combination of manual traversal and boundary node traversal.

In addition to this the current logic does not take care of various edge cases, for example container sets.

It was clear some rethinking needed to be done such that:
a) edge cases were taken care of.
b) retry logic was simpler to understand and hopefully fix if any further bug were to arise.

### Modifications

This is a complete rewrite of the `FormulateRetryWorkflow` function.

### Verification

Extensive testing performed manually to verify that behaviour is as expected.

### CODE REVIEW

**Comment by Joibel** [comment]
This might superceed #12105

File: workflow/util/util_test.go [hunk 1]
```diff
@@ -2092,18 +2014,1794 @@ func TestRetryWorkflowWithNestedDAGsWithSuspendNodes(t *testing.T) {
...
+func TestNestedDAG(t *testing.T) {
+	require := require.New(t)
+	wf := wfv1.MustUnmarshalWorkflow(nestedDAG)
+
+	newWf, podsToDelete, err := FormulateRetryWorkflow(context.Background(), wf, true, "id=dag-nested-zxlc2-744943701", []string{})
+	require.NoError(err)
+	_ = newWf
+	_ = podsToDelete
+
+}
```
Comment (isubasinghe) [inline workflow/util/util_test.go:3807]: These should actually test for individual node status as well.

File: workflow/util/util.go [hunk 2]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+func consumePod(n *node, resetFunc resetFn, addToDelete deleteFn) (*node, error) {
+	// this sets to reset but resets are overridden by deletes in the final FormulateRetryWorkflow logic.
+	curr, err := consumeTill(n, getTillFnNodeType(wfv1.NodeTypePod), resetFunc)
+	if err != nil {
+		return nil, err
+	}
+	addToDelete(curr.n.ID, true)
+	children := getChildren(curr)
+	for childID := range children {
+		addToDelete(childID, true)
```
Comment (Joibel) [inline workflow/util/util.go:1028]: This recurses one child only, which is surprising... one child might be correct today but it might be a fun source of orphans in the future. This is not a sentence I expected to write.

Ah, I suspect that new line 1346 prevents us keeping children of deleted nodes anyway, so this whole child deletion loop could be removed instead.

File: workflow/util/util.go [hunk 3]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+	}
+
+	addToReset := func(nodeID string, addToNode bool) {
+		if nodeID == toNode && !addToNode {
+			return
+		}
+		nodesToReset[nodeID] = true
+	}
+
+	addToDelete := func(nodeID string, addToNode bool) {
```
Comment (Joibel) [inline workflow/util/util.go:1062]: `addToNode` is always `true`, so simplify

File: workflow/util/util.go [hunk 4]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+
+	if err != nil {
+		return nil, nil, err
+	}
+	l := len(nodes)
+	if l <= 0 {
+		return nodesToReset, nodesToDelete, nil
+	}
+
+	addToReset := func(nodeID string, addToNode bool) {
```
Comment (Joibel) [inline workflow/util/util.go:1055]: `addToNode` is always `false`, so simplify

File: workflow/util/util.go [hunk 5]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+	return res
+}
+func shouldRetryFailedType(nodeTyp wfv1.NodeType) bool {
+	if nodeTyp == wfv1.NodeTypePod || nodeTyp == wfv1.NodeTypeContainer {
+		return true
+	}
+	return false
+}
+
+// dagSortedNodes sorts the nodes based on the order they were created, omits onExitNode
```
Comment (Joibel) [inline workflow/util/util.go:1171]: Technically not in creation order as children are all appended batched without regard to order.

If `task-1` finishes later than `task-2` then task-1's children may  get created later than task-2's but that won't be reflected here.

File: workflow/util/util.go [hunk 6]
```diff
@@ -853,178 +1218,152 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
...
+		toReset = setUnion(toReset, pathToReset)
+		toDelete = setUnion(toDelete, pathToDelete)
 	}
 
-	onExitNodeName := wf.ObjectMeta.Name + ".onExit"
-	// Get all children of nodes that match filter
-	nodeIDsToReset, err := getNodeIDsToReset(restartSuccessful, nodeFieldSelector, wf.Status.Nodes)
-	if err != nil {
-		return nil, nil, err
+	for k := range toReset {
```
Comment (Joibel) [inline workflow/util/util.go:1288]: Swap `k` for `nodeName`. Short variable names for array indices are OK, but here it's a map key and more confusing if not named for what it is.

File: workflow/util/util.go [hunk 7]
```diff
@@ -853,178 +1218,152 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
...
-				deletedNodes[node.ID] = true
+				queue.Remove(currNode)
 			}
-			// do not add this status to the node. pretend as if this node never existed.
-		default:
-			// Do not allow retry of workflows with pods in Running/Pending phase
-			return nil, nil, errors.InternalErrorf("Workflow cannot be retried with node %s in %s phase", node.Name, node.Phase)
+		}
+		if n.Name == wf.Name && !shouldRetryFailedType(n.Type) {
+			newWf.Status.Nodes.Set(id, resetNode(*n.DeepCopy()))
```
Comment (Joibel) [inline workflow/util/util.go:1335]: Do we need this `DeepCopy`, it seems to me that `resetNode` will cause a copy?

File: workflow/util/util_test.go [hunk 8]
```diff
@@ -2092,18 +2014,1794 @@ func TestRetryWorkflowWithNestedDAGsWithSuspendNodes(t *testing.T) {
...
+  startedAt: "2024-09-19T02:43:30Z"
+  taskResultsCompletionStatus:
+    steps-4k5vn-1044844302: true
+    steps-4k5vn-2627784879: true
+    steps-4k5vn-2644562498: true
+
+`
+
+func TestStepsRetryWorkflow(t *testing.T) {
+	assert := assert.New(t)
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util_test.go:2253]: Ah I was waiting for an opinion on https://github.com/argoproj/argo-workflows/pull/13734#discussion_r1808224838. I plan to use these when I fill in the tests. 

I was unsure if it was worth the effort of these tests. I wrote these to test edge conditions and make sure they worked as expected. They will prevent regressions though, so I will add them.

File: workflow/util/util_test.go [hunk 9]
```diff
@@ -1026,51 +1025,6 @@ func TestRetryExitHandler(t *testing.T) {
 func TestFormulateRetryWorkflow(t *testing.T) {
 	ctx := context.Background()
 	wfClient := argofake.NewSimpleClientset().ArgoprojV1alpha1().Workflows("my-ns")
-	createdTime := metav1.Time{Time: time.Now().Add(-1 * time.Second).UTC()}
-	finishedTime := metav1.Time{Time: createdTime.Add(time.Second * 2)}
-	t.Run("Steps", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util_test.go:1031]: I will check those now, the logic differed quite a lot. This sets nodes to Running I believe, whereas I delete them. 
This is needed to actually get a pod allocated, this why sometimes when  you hit retry, a workflow can be stuck on Running.

File: workflow/util/util.go [hunk 10]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+type parentMap map[string]*wfv1.NodeStatus
+
+type node struct {
+	n        *wfv1.NodeStatus
+	parent   *node
+	children []*node
+}
+
+func newWorkflowsDag(wf *wfv1.Workflow) ([]*node, error) {
+
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util.go:835]: I love my blank lines :)

File: workflow/util/util.go [hunk 11]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+	}
+
+	slices.Reverse(reverseNodes)
+	return reverseNodes, nil
+}
+
+func getChildren(n *node) map[string]bool {
+	children := make(map[string]bool)
+	queue := list.New()
+	queue.PushBack(n)
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util.go:924]: No real reason, I prefer non-recursive functions in non-functional languages, mostly owing to eliminating a source of stack overflows. 

I don't think we would ever stack overflow in practice here,.
I think Go's tail call optimisation isn't great as well so didn't want to rely upon that.

File: workflow/util/util.go (hunk 5 above)
Comment (isubasinghe) [reply workflow/util/util.go:1171]: Yeah its more, topological sorted order. Will change that.

File: workflow/util/util.go (hunk 7 above)
Comment (isubasinghe) [reply workflow/util/util.go:1335]: It will cause a copy but only a shallow one. This isn't great for nodes that will mutate over time. 
A reset node will mutate so it does need a deep copy.  There probably isn't pointers shared across multiple nodes, but its better to be safe than introduce some weird aliasing bug. 

I was actually missing one so this comment helped me catch it, so thanks.

File: workflow/util/util_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util_test.go:1031]: It's not really testable easily, we cannot establish a dag for these status fields.  The root node is not present. 
This will error out complaining that the root node couldn't be found.

File: workflow/util/util.go (hunk 2 above)
Comment (isubasinghe) [reply workflow/util/util.go:1022]: > This recurses one child only 

I don't think so ? We get all children of curr. The getChildren does a BFS to obtain all the children 

> Ah, I suspect that new line 1346 prevents us keeping children of deleted nodes anyway, so this whole child deletion loop could be removed instead.

Yes but we shouldn't rely on that kind of inference in my opinion. For all other executable node types, we need to track the deletion id so that we can delete the pods. We shouldn't diverge from this logic just because it happens not  to be needed for containersets.

File: workflow/util/util.go (hunk 4 above)
Comment (isubasinghe) [reply workflow/util/util.go:1049]: It is actually safe to get rid  of the `addToNode` arg here, the reason being the deletion overrides the resetting. 
Maybe a comment  explaining so and removing this arg is best.

File: workflow/util/util.go [hunk 12]
```diff
@@ -799,50 +788,426 @@ func deletePodNodeDuringRetryWorkflow(wf *wfv1.Workflow, node wfv1.NodeStatus, d
...
+			if shouldReset {
+				resetFunc(curr.n.ID, true)
+			}
+			return curr, nil
+		}
+		curr = curr.parent
+	}
+}
+
+func getTillBoundaryFn(boundaryID string) tillFn {
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/util/util.go:962]: done!

File: workflow/util/util.go [hunk 13]
```diff
@@ -880,151 +858,437 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
...
-				log.Debugf("Deleted %s node %s since it's not a group node", node.Name, string(node.Phase))
-				deletedPods, podsToDelete = deletePodNodeDuringRetryWorkflow(wf, node, deletedPods, podsToDelete)
-				log.Debugf("Deleted pod node: %s", node.Name)
-				deletedNodes[node.ID] = true
-			}
-			// do not add this status to the node. pretend as if this node never existed.
-		default:
-			// Do not allow retry of workflows with pods in Running/Pending phase
-			return nil, nil, errors.InternalErrorf("Workflow cannot be retried with node %s in %s phase", node.Name, node.Phase)
+	deleteNodes, err := getNodeIDsToResetNoChildren(restartSuccessful, nodeFieldSelector, wf.Status.Nodes)
```
Comment (gzbvas) [inline workflow/util/util.go]: Would I have to set the --restart-successful in order to identify the nodes that need to be deleted, even if I only wish to retry a specific failed node?

Comment (isubasinghe) [reply workflow/util/util.go]: No this wouldn't be the case

File: workflow/util/util.go (hunk 2 above)
Comment (Joibel) [reply workflow/util/util.go]: Ok, I follow and agree now. Thanks.

**Review by Joibel — APPROVED** [review]
LGTM.

**Comment by shuangkun** [comment]
Could this PR be cherry-picked into the 3.6.5 branch? Some customers have inquired about it, but I haven't found any prior discussions related to the 3.6.x patches. Thanks @Joibel

**Comment by Joibel** [comment]
I've created a patch releases discussion #14259.

I haven't backported this yet, it feels like #14124 should be addressed before this is backported otherwise we're knowingly introducing a regression.
